### DISCUSSION TITLE
parsing graphql query

### DISCUSSION THREAD

**Discussion by someone-1** [graphql-java/graphql-java#3337] [Q&A]
I need to parse graphql query like 
```
query BookById {
          books {
                    id
                    name
                    pageCount
           }
}
```
and convert it to a java object 
any suggestion on which class/method I can refer for this

**Comment by bbakerman** [comment]
See `graphql.parser.Parser` with stacks of examples in the test code.  This is the raw AST document.

There is also a higher level mechanism called `graphql.ParseAndValidate` which as its name suggests can parse and validate

> and convert it to a java object

What would you do with this Java object?  Is it mean to represent a query or represent what would come back from a query?

eg are you after code generation here ?

**Reply by someone-1** [reply]
thanks will check this.

>What would you do with this Java object? Is it mean to represent a query or represent what would come back from a query?

I am identifying the graphql query type, operation-name and selection set for a given query

**Reply by bbakerman** [reply]
Ahh then you only need `Parser` - the AST nodes can tell you all that.

Selection set is tricky because of fragments.  the above query can be written as follows and executes as the same fragment

```
query BookById {
          books {
                    id
                    ... NameParts
                    ... on Book {
                       pageCount
                    }
           }
}

fragment NameParts on Book {
   name
}
```


So depending on how complex you will allow your query sub selections to be, collecting this is trickier.
